packages/dds/map/src/test/directoryOracle.ts

```diff
@@ -19,37 +17,58 @@ import type {
  * @internal
  */
 export class SharedDirectoryOracle {
-	private readonly model = new Map<string, unknown>();
+	// Model updated via valueChanged events
+	private readonly modelFromValueChanged = new Map<string, unknown>();
+
+	// Model updated via containedValueChanged events
+	private readonly modelFromContainedValueChanged = new Map<string, unknown>();
+
+	// Track all directories we've attached listeners to for proper cleanup
+	private readonly attachedDirectories = new Set<IDirectory>();
 
 	public constructor(private readonly sharedDir: ISharedDirectory) {
-		// Capture initial state BEFORE attaching event listeners
-		// to avoid double-counting any events that might fire during initialization
-		this.captureInitialSnapshot(sharedDir);
+		this.snapshotCurrentState(this.sharedDir);
 
-		// Now attach event listeners for future changes
+		// valueChanged fires globally on the root for ALL changes anywhere in the tree
+		// Only needs one listener on the root (includes 'path' field to indicate location)
 		this.sharedDir.on("valueChanged", this.onValueChanged);
 		this.sharedDir.on("clear", this.onClear);
 		this.sharedDir.on("subDirectoryCreated", this.onSubDirCreated);
 		this.sharedDir.on("subDirectoryDeleted", this.onSubDirDeleted);
-		this.sharedDir.on("containedValueChanged", this.onContainedValueChanged);
+
+		// containedValueChanged fires locally on each directory for its own changes
+		// Needs separate listeners on every subdirectory (no 'path' field, uses target.absolutePath)
+		// disposed/undisposed events are also attached to subdirectories for rollback support
+		this.attachToAllDirectories(sharedDir);
 	}
 
-	private captureInitialSnapshot(dir: IDirectory): void {
-		const { absolutePath } = dir;
+	private attachToAllDirectories(dir: IDirectory): void {
+		// Track this directory for cleanup
+		this.attachedDirectories.add(dir);
 
-		// Capture keys
-		for (const [key, value] of dir.entries()) {
-			const pathKey = absolutePath === "/" ? `/${key}` : `${absolutePath}/${key}`;
-			this.model.set(pathKey, value);
-		}
+		// Attach containedValueChanged listener to this directory
+		dir.on("containedValueChanged", this.onContainedValueChanged);
```

Comment:
Subdirectories probably also need to register for subDirectoryCreated/Deleted to observe further descendent subdirectories being created/deleted?